### ISSUE TITLE
Custom validation/linting rules

### ISSUE THREAD

**Issue by jamieshorten** [accordproject/concerto#814]
# Feature Request 🛍️

Create a linter to allow users to specify rules for how their Concerto models are structured

## Use Case
It would be useful to  have a tool to allow users to do some / all of the following:
1. Specify the naming of declarations. E.g. all names of scalars should be in camel case.
2. Specify the naming of properties, enum cases e.t.c
3. Specify which language features can be used. E.g. disallow maps, disallow forward references in regex validators.
4. Enforce the use of certain features. E.g. all string properties should have a length validator.
5. Enforce the use of `@Term` decorators on all declarations and properties e.t.c
6. All concepts in a namespace should extend a given concept
7. All concepts in a namespace must have unique names across multiple namespaces

## Context
1. We have tools and applications that support different subsets of the language. It would be useful to enforce a common subset of functionality.
2. It would be useful to enforce a style guide around naming and best practices. 
3. It would be useful to enforce these rules at the level of the ModelManager?

**Comment by vr-varad** [comment]
Hi @sanketshevkar, I'm eager to work on the pull request. Could you please assign it to me? Thanks!

**Comment by sanketshevkar** [comment]
hey @vr-varad 
This issue is one of the ideas for GSoC'24. We suggest you to visit https://summerofcode.withgoogle.com/how-it-works
to get more details.

**Comment by sanketshevkar** [comment]
Please use this issue for discussion or doubts regarding this project idea.

**Comment by sanketshevkar** [comment]
Yes! Conversion of Concerto model to JSON AST already exists. But the linting part is missing.

**Comment by subhajit20** [comment]
> Yes! Conversion of Concerto model to JSON AST already exists. But the linting part is missing.

Can you point out where the `AST` conversation code is written ?

**Comment by sanketshevkar** [comment]
You need to add file to the [modelManager](https://github.com/accordproject/concerto/blob/0fca1e8bc3011bc265074fbda67cbd4dbcb1df26/packages/concerto-core/lib/basemodelmanager.js#L208), that will give you a [modelFile](https://github.com/accordproject/concerto/blob/main/packages/concerto-core/lib/introspect/modelfile.js) object, which has a [getAST](https://github.com/accordproject/concerto/blob/0fca1e8bc3011bc265074fbda67cbd4dbcb1df26/packages/concerto-core/lib/introspect/modelfile.js#L672) method, which would return the JSON AST.

@dselman is there any better way for this?

**Comment by dselman** [comment]
No, that's the way to do it if you want to ensure the AST is valid and consistent (references resolve).

**Comment by dselman** [comment]
Here's a tool for Open API definition linting that might spark some ideas:
https://stoplight.io/open-source/spectral

**Comment by subhajit20** [comment]
> We can use Spectral as the framework for defining our own rules over the Concerto AST (JSON): https://github.com/stoplightio/spectral
> 
> Here are two sample rules:
> 
> ```
> rules:
>   declaration-pascal-case:
>     description: Declaration names should use Pascal case.
>     message: "Declaration '{{value}}' should use Pascal case."
>     severity: warn
>     given: $.declarations[*].name
>     then:
>       function: casing
>       functionOptions:
>         type: pascal
>   enum-value-macro-case:
>     description: Enum values should use Macro case.
>     message: "Enum '{{value}}' should use Macro case."
>     severity: warn
>     given: $.declarations[?(@.$class=="concerto.metamodel@1.0.0.EnumDeclaration")].properties[*].name
>     then:
>       function: casing
>       functionOptions:
>         type: macro
> ```
> 
> When used with this CTO:
> 
> ```
> namespace test@1.0.0
> 
> enum Gender {
>     o Male
>     o Female
>     o Other
> }
> 
> concept person {
>     o String name
> }
> ```
> 
> To create the AST:
> 
> ```
> concerto parse --model test.cto --excludeLineLocations --output test.json
> ```
> 
> test.json:
> 
> ```json
> {
>   "$class": "concerto.metamodel@1.0.0.Model",
>   "decorators": [],
>   "namespace": "test@1.0.0",
>   "imports": [],
>   "declarations": [
>     {
>       "$class": "concerto.metamodel@1.0.0.EnumDeclaration",
>       "name": "Gender",
>       "properties": [
>         {
>           "$class": "concerto.metamodel@1.0.0.EnumProperty",
>           "name": "Male"
>         },
>         {
>           "$class": "concerto.metamodel@1.0.0.EnumProperty",
>           "name": "Female"
>         },
>         {
>           "$class": "concerto.metamodel@1.0.0.EnumProperty",
>           "name": "Other"
>         }
>       ]
>     },
>     {
>       "$class": "concerto.metamodel@1.0.0.ConceptDeclaration",
>       "name": "person",
>       "isAbstract": false,
>       "properties": [
>         {
>           "$class": "concerto.metamodel@1.0.0.StringProperty",
>           "name": "name",
>           "isArray": false,
>           "isOptional": false
>         }
>       ]
>     }
>   ]
> }
> ```
> 
> Produces the output:
> 
> ```
> spectral lint test.json
> 
> /Users/dan.selman/dev/concerto-spectral/test.json
>  13:19  warning  enum-value-macro-case    Enum 'Male' should use Macro case.            declarations[0].properties[0].name
>  17:19  warning  enum-value-macro-case    Enum 'Female' should use Macro case.          declarations[0].properties[1].name
>  21:19  warning  enum-value-macro-case    Enum 'Other' should use Macro case.           declarations[0].properties[2].name
>  27:15  warning  declaration-pascal-case  Declaration 'person' should use Pascal case.  declarations[1].name
> ```

I have gone through the readme section and I am having one question that - 
Do we need to create a separate package for example `concerto-lint` which will use `spectral` framework where we can define our rulesets in the form of `JSON` or. `YML`  and then we will pass the folder name via command line like `concerto-lint test.json` and it will validate all the json file and log all the modifications needed for the cto model

**Comment by apoorvsxna** [comment]
Hi. I've been looking into this project's requirements and am interested in working on it. This is my current understanding of how the linter would work for a general user, please correct me if I'm wrong-

1. Create concerto model.
2. Open up terminal and parse model using the linter cli.
3. Linter converts model to AST, and checks each element for rule violations (pre-specified spectral rules)
4. Linter reports rule violating elements as output.

**Comment by sanketshevkar** [comment]
Yes, that's correct! For the following point
> Linter converts model to AST, and checks each element for rule violations (pre-specified spectral rules)

We already have methods in Concerto that can get you the AST, rule checking is something that you'll have to build.

**Comment by sanketshevkar** [comment]
> Do we need to create a separate package for example concerto-lint which will use spectral framework where we can define our rulesets in the form of JSON or. YML and then we will pass the folder name via command line like concerto-lint test.json and it will validate all the json file and log all the modifications needed for the cto model ?

Something like that. It sounds like a good approach to start with.

**Comment by subhajit20** [comment]
> > Do we need to create a separate package for example concerto-lint which will use spectral framework where we can define our rulesets in the form of JSON or. YML and then we will pass the folder name via command line like concerto-lint test.json and it will validate all the json file and log all the modifications needed for the cto model ?
> 
> Something like that. It seems like a good approach to start with. I haven't played much around with spectral yet, so please take this with a grain of salt. On a high level it seems like a good approach to me.

Hey I have tried both options like cli and javascript ruleset both worked fine 

`async function lintConcertoModel() {
    const data = fs.readFileSync('./test.cto','utf8');
    const model_manager = new ModelManager()
    const modelFile = model_manager.addCTOModel(data,'test.cto');
    const ast = modelFile.getAst();

    const spectral = new Spectral();
    spectral.setRuleset({
        rules: {
            "concept-name-capitalized": {
              description:"Concept names should start with a capital letter.",
              given: "$.declarations[*].name",
              message: "Declaration '{{value}}' should use Pascal case.",
              then: {
                field:"name",
                function: pattern,
                functionOptions:{
                    match:"^[A-Z]"
                }
              },
            },
        }
    });

    spectral.run(ast).then((e)=>{
        console.log(e)
    })
    .catch((err)=>{

        console.log(err)
    })
  
    // return results;
  }
`

- Output
`[
  {
    code: 'concept-name-capitalized',
    message: "Declaration 'person' should use Pascal case.",
    path: [ 'declarations', '0', 'name' ],
    severity: 1,
    range: { start: [Object], end: [Object] }
  },
  {
    code: 'concept-name-capitalized',
    message: "Declaration 'subhajit' should use Pascal case.",
    path: [ 'declarations', '1', 'name' ],
    severity: 1,
    range: { start: [Object], end: [Object] }
  }
]`

**Comment by kairveeehh** [comment]
hey @dselman  as this issue had been under gsoc'24 but still seems to be open , could you let me know about it if its available to work and what were the complications faced with the previous flow that was listed in the comments? 
thanks

**Comment by dselman** [comment]
> hey @dselman as this issue had been under gsoc'24 but still seems to be open , could you let me know about it if its available to work and what were the complications faced with the previous flow that was listed in the comments? thanks

Yes, the issue is still open. There wasn't any major issue with the approach described, it was just deprioritised for GSoC. I still think it would be a valuable contribution, and by leveraging Spectral should not be too costly or difficult to implement.

**Comment by Ahmed-Gaper** [comment]
Hi @dselman 
I come with some potential new rules for the Concerto linter after testing them in actual projects. Can you double-check these to see if I can add them to my Concerto linter proposal? Let me know if any might cause problems, or if there’s something I overlooked

- All string properties must have a default value (use "" if no value is needed to avoid runtime error).

- If a property has no default value, must use the @Required decorator.

- Disallow maps. Disallow forward references in regex validators. Arrays are allowed but must set max size.

- Regex validators should use predefined patterns (no inline regex), to avoid Complex regex and Ensure the same regex is reused everywhere.

- All transactions should reference at least one participant.

- All events should include a DateTime property named "timestamp".

- UPPER_SNAKE_CASE for Enum Constants. Each constant must have a @Term decorator.

**Comment by sanketshevkar** [comment]
Hi @Ahmed-Gaper 
> All string properties must have a default value (use "" if no value is needed to avoid runtime error).
If a property has no default value, must use the @Required decorator.

For these two please look into language specs, we have optional and default meta-properties for properties and they can be used. https://concerto.accordproject.org/docs/design/specification/model-properties#meta-properties

> Regex validators should use predefined patterns (no inline regex), to avoid Complex regex and Ensure the same regex is reused everywhere.

Do you mean enforcing scalars for regex?

Rest sound fine to me.

**Comment by Ahmed-Gaper** [comment]
@sanketshevkar 

Regarding regex scalars, they don't support regex declarations, right? I suggest adding that feature so we can reuse one regex with different properties without repeating ourselves, without enforcing this as a rule.

And for the first point: do you mean this linting rule be like that strings should get a `default `meta-property when not marked as `optional `to avoid runtime errors, correct?

**Comment by sanketshevkar** [comment]
> And for the first point: do you mean this linting rule be like that strings should get a default meta-property when not marked as optional to avoid runtime errors, correct?

Yes, `@Required` decorator doesn't have any logic associated with it the model validator. `optional` meta-property has.

> Regarding regex scalars, they don't support regex declarations, right? I suggest adding that feature so we can reuse one regex with different properties without repeating ourselves, without enforcing this as a rule.

Can you please model this, I'm not able to form this in my mind. 
> Regarding regex scalars, they don't support regex declarations, right?

**Comment by Ahmed-Gaper** [comment]
> Yes, `@Required` decorator doesn't have any logic associated with it the model validator. `optional` meta-property has. Decorators are just labels, concerto libraries themselves don't have any logic associated with any label. Its mostly used by other systems which would want to transform a model. Eg: if you want to render a model in frontend, if you want to transform your model to any other schema and want a specific transformation based on that decorator label.

I understood your first point and will follow it. 

> Can you please model this, I'm not able to form this in my mind.

After working through this, I noticed the scalar we currently use already does what I was trying to propose. Your insight helped me connect the dots—thank you so much!
